### DISCUSSION TITLE
.NET Core 3.1 - support for including CefSharp in library?

### DISCUSSION THREAD

**Discussion by omaristalis** [cefsharp/CefSharp#3401] [Q&A]
Hi there,

I thought I'd post here instead of raising an issue as this might be by design, or I might be missing something.

I created a .net core library project using CefSharp.WinForms, and everything seems to be well - I even multi-targetted that library for net472 and netcoreapp3.1, and I could use that everywhere - .net core app, framework app and even interopped into a native C++/CLI app.

However, when I tried converting the NET Core 3.1 library to use CefSharp.WinForms.NETCore, I ran into problems.  The output of the build would only include 2 files in the x64/x64 native paths - CefSharp.dll and ijwhost.dll.  Everything else is missing, include cefsharp.core.dll (which is the error that you get when you try and use it).

I created a new project from scratch which targetted only netcore 3.1, and had the same issue.

The behaviour is the same when the platform is AnyCPU or x64.

The project file is pretty simple (note I tried every combination of platforms, each on their own and both included as below):

```
<Project Sdk="Microsoft.NET.Sdk.WindowsDesktop">

  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <Platforms>AnyCPU;x64</Platforms>
    <UseWindowsForms>true</UseWindowsForms>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="CefSharp.WinForms.NETCore" Version="87.1.132" />
    <PackageReference Include="CefSharp.Common.NETCore" Version="87.1.132" />
    <PackageReference Include="System.Text.Json" Version="5.0.0" />
    <PackageReference Include="Microsoft.AspNetCore.WebUtilities" Version="2.2.0" />
    <PackageReference Include="System.IdentityModel.Tokens.Jwt" Version="6.8.0" />
    <PackageReference Include="System.Resources.Extensions" Version="5.0.0" />
  </ItemGroup>

  <ItemGroup>
    <Compile Update="BrowserForm.cs">
      <SubType>Form</SubType>
    </Compile>
    <Compile Update="LogoutForm.cs">
      <SubType>Form</SubType>
    </Compile>
  </ItemGroup>

</Project>
```

I created a new net core 3.1 winforms application and copied the exact same code over (the only change was to add a button to the main app form to call my code), and it seems to behave - all of the binaries are copied over.

Is including the NET Core version of cefsharp in a library not supported?

Thanks in advance!

**Comment by amaitland** [comment]
I'd suggest adding a RuntimeIdentifier

Please read
- https://github.com/cefsharp/CefSharp/blob/cefsharp/87/NuGet/PackageReference/Readme.txt#L10
- https://github.com/cefsharp/CefSharp/issues/3284#issuecomment-772132523

Add a RuntimeIdentifier, for dev purposes I use something like https://github.com/cefsharp/CefSharp/issues/3197#issuecomment-723537000

For reference this is a limitation of `Nuget` see https://github.com/NuGet/Home/issues/6083#issuecomment-340670584

> Is including the NET Core version of cefsharp in a library not supported?

It's actually not something I've had time to test, the references should be transitive and should flow through to the consuming library without issue. As this is the very first release there's likely improvements that can be made.

**Reply by amaitland** [reply]
> For reference this is a limitation of `Nuget` see [NuGet/Home#6083 (comment)](https://github.com/NuGet/Home/issues/6083#issuecomment-340670584)

Just to clarify on this. When no `RuntimeIdentifier` is specified I've added some hacks in an attempt to make it work without one. 

- The `.targets` file copies the `CEF` resources into the relevant subfolders see
https://github.com/cefsharp/CefSharp/blob/cefsharp/87/NuGet/PackageReference/CefSharp.Common.NETCore.targets#L40
- The `ModuleInitializer` attempts to locate the correct folder and load `libcef.dll` and subsequently set the paths see https://github.com/cefsharp/CefSharp/blob/cefsharp/87/CefSharp.Core/Initializer.cs#L28

This really is a hack and whilst should work in most cases, isn't guaranteed, specifying a `RuntimeIdentifier` is the recommended approach.
